base/applications/cmdutils/taskkill/taskkill.c

```diff
@@ -249,57 +356,90 @@ static int send_close_messages(void)
                 continue;
             }
 
+            // Send close messages to child first
+            if (kill_child_processes)
+            {
+                send_close_messages_tree(pid);
+            }
+
+            get_process_name_from_pid(pid, process_name, MAX_PATH);
             EnumWindows(pid_enum_proc, (LPARAM)&info);
             if (info.found)
-                taskkill_message_printfW(STRING_CLOSE_PID_SEARCH, pid);
-            else
             {
-                taskkill_message_printfW(STRING_SEARCH_FAILED, task_list[i]);
-                status_code = 128;
+                if (is_numeric)
+                {
+                    taskkill_message_printfW(STRING_CLOSE_PID_SEARCH, pid);
+                }
+                else
+                {
+                    taskkill_message_printfW(STRING_CLOSE_PROC_SRCH, process_name, pid);
+                }
             }
         }
-        else
-        {
-            DWORD index;
-            BOOL found_process = FALSE;
 
-            for (index = 0; index < pid_list_size; index++)
+        HeapFree(GetProcessHeap(), 0, pkill_list);
```

Comment:
(If you agree with the comment above concerning `pkill_list`, then you could move this `HeapFree()` call outside of the for-loop.)